app/vmalert/rule/alerting.go

```diff
@@ -50,11 +51,26 @@ type AlertingRule struct {
 }
 
 type alertingRuleMetrics struct {
-	errors        *utils.Counter
-	pending       *utils.Gauge
-	active        *utils.Gauge
-	samples       *utils.Gauge
-	seriesFetched *utils.Gauge
+	set *metrics.Set
+
+	errors        *metrics.Counter
+	pending       *metrics.Gauge
+	active        *metrics.Gauge
+	samples       *metrics.Gauge
+	seriesFetched *metrics.Gauge
+}
+
+func newAlertingRuleMetrics() *alertingRuleMetrics {
+	set := metrics.NewSet()
+	metrics.RegisterSet(set)
```

Comment:
>expose group metrics only once group is started

Little inconsistency between group and rule metricsSet register, so there will be rule metrics but without group metrics before the group starts, can we sync them somehow?

Reply:
@Haleygo Seems like it might make it complicated if we'd try to sync the rules because of in-place updates for groups and rules.
With an updated logic each group will have its own `metrics.Set`. When a new group is created after parsing config and new rules are created these rules will be linked to a new `metrics.Set`.
Now when `updateWith` will be used to update existing `Group` in place rules will need to somehow re-register metrics to `metrics.Set` of existing `Group`.
There are a few ways to fix this, but it seems like it will make the code more complicated for the sake of exposing rule metrics at the same time as the group starts.
Maybe you have some ideas on how it's possible to sync these without making it more complicated?

Reply:
Currently this code produces metrics duplicates.

 Just run a small test locally with 2 groups. If 1 alerting rule at group 1 modified, alerting rules from group 2 will be duplicated. It leads to endless response size grows and OOM with-in some time.

 I have two options:
1) register alert metrics when group actually starts
2) never de-register metrics and use GetOrCreate methods.